onDragEnd event, when user finished selecting value

vitalybaev opened the pull request:
This improves #32 PR and relates to #30 
Event fired in these cases:
1. When a user ends dragging (mouse/touch)
2. Keyboard key ups

But I may forget some cases.
I saw, that you prefer `onDragEnd` name. But I guess the event should be fired also when a user makes a single click. And in this case, I think this event should have a different name.
What do you think?

- [x] TypeScript types
- [x] Flow types
- [x] Readme

tajo commented on src/Range.tsx:
```diff
@@ -16,7 +16,13 @@ import {
+interface IState {
+  draggedThumbIndex: number;
+  thumbZIndexes: number[];
+  isChanged?: boolean;
```
Why is there a need to detect this? I would expect `onDragEnd` to always fire.

If really needed, `isChanged: false` can be used as the initial state and no need to define `IState` interface.

vitalybaev replied:
The point is that not every keyup event should fire `onFinalUpdate` event e.g when we pressed the `Tab` button.
But you are right that it could be in an initial state without defining an interface

tajo replied:
Oh, I get it now. Well, it would be better to just filter key events the same way as it is done in `onKeyDown` (ideally, extract those filters/conditions into so it's in a single place). No need to introduce an another state, right?

vitalybaev replied:
I've found some issue with this case and offer to switch back to the state:
When you press some arrow key, thumb moved, value changed. But if you press another key, e.g. `Tab` without raising arrow key, we lose a focus and `onFinalChange` won't be fired.

vitalybaev replied:
Fixed this and added test.

tajo commented:
> I saw, that you prefer `onDragEnd` name. But I guess the event should be fired also when a user makes a single click. And in this case, I think this event should have a different name.
> What do you think?

Hm, how about `onFinalUpdate`?

tajo commented on README.md:
````diff
@@ -157,6 +157,14 @@ onChange: (values: number[]) => void;
+onDragEnd: (values: number[]) => void;
+```
+
+Called when a drag is finished, provides current `values`.
````
should mention the keyboard

tajo commented:
Thanks! Looking good. 

Before the merge, there should be an e2e test for both dnd and keyboard interaction.

vitalybaev commented:
> Hm, how about onFinalUpdate?
Sounds good

vitalybaev commented:
Updates:
1. Renamed to `onFinalChange`
2. Updated readme, ts and flow types
3. Removed `isChanged` and `IState`.
4. Added one e2e test. Tell me if I do it the right way

tajo commented:
The PR looks great; however, it seems that the keyboard interaction doesn't work in the [example](https://deploy-preview-36--react-range.netlify.com/?path=/story/range--onfinalchange-event).

It would be great to add a test for the keyboard too.

vitalybaev commented:
> it seems that the keyboard interaction doesn't work

I'll take a look into this.

> It would be great to add a test for the keyboard too.

Sure. I've just written the test to ensure that I do it the right way.

vitalybaev commented:
> it seems that the keyboard interaction doesn't work

Oh, my bad. Forgot about event pooling in React. Fixed it now.

And added e2e keyboard test.

tajo commented:
hm, the mouse test is now failing, might be flaky?

vitalybaev commented:
Maybe, I've executed tests locally and it was green.

vitalybaev commented:
> hm, the mouse test is now failing, might be flaky?

Yep, it was some kind of magic. I've pushed fixes for a simultaneous keypress while dragging with a keyboard.
As for me, now it works as expected, and looks final.

tajo commented on src/Range.tsx:
```diff
@@ -258,12 +266,30 @@ class Range extends React.Component<IProps> {
   onKeyUp = (e: React.KeyboardEvent) => {
-    this.setState({ draggedThumbIndex: -1 });
+    const { isChanged } = this.state;
+    console.log('keyup', 'isChanged', isChanged);
```
```suggestion
```

vitalybaev replied:
Oops. Usually, linters don't let me this into a repository.

vitalybaev commented:
Added a little fix - if while dragging thumb via keyboard we press any other key - `keyUp` won't be fired so `isChanged` won't be false when calling `onFinalChange`

tajo commented:
Thanks!

```
+ react-range@1.4.0
```
